### PR TITLE
SI-9246 *OrElse should have a by-name parameter

### PR DESCRIPTION
**Pull request by 193s**
reopen #4279 :gem: 
Sorry for the trouble.

### CODE REVIEW

**Comment by adriaanm** [comment]
No worries, @193s! Thank you for your contribution. It looks like the publish-core task discovered an issue with this submission:

```
Compiling 314 files to /home/jenkins/workspace/scala-2.11.x-validate-publish-core/build/locker/classes/compiler
/home/jenkins/workspace/scala-2.11.x-validate-publish-core/src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:616: warning: matches with two cases or fewer are emitted using if-then-else instead of switch
              (argsSize : @switch) match {
                           ^
/home/jenkins/workspace/scala-2.11.x-validate-publish-core/src/compiler/scala/tools/reflect/WrappedProperties.scala:23: error: method propOrElse overrides nothing.
Note: the super classes of trait WrappedProperties contain the following, non final members named propOrElse:
def propOrElse(name: String,alt: => String): String
  override def propOrElse(name: String, alt: String) = wrap(super.propOrElse(name, alt)) getOrElse alt
               ^
/home/jenkins/workspace/scala-2.11.x-validate-publish-core/src/compiler/scala/tools/reflect/WrappedProperties.scala:26: error: method envOrElse overrides nothing.
Note: the super classes of trait WrappedProperties contain the following, non final members named envOrElse:
def envOrElse(name: String,alt: => String): String
  override def envOrElse(name: String, alt: String)  = wrap(super.envOrElse(name, alt)) getOrElse alt
               ^
one warning found
two errors found
```

**Comment by 193s** [comment]
I fixed! Thank you, @adriaanm!

```
BUILD SUCCESSFUL
Total time: 11 minutes 32 seconds
```

**Comment by adriaanm** [comment]
Great! You may ignore the IDE failure (#4287 has been merged to ignore it). The test failure is because this PR introduces a binary incompatible change in an API used by partest, which is compiled separately: scala/scala-partest.

File: src/library/scala/util/Properties.scala [hunk 1]
```diff
@@ -57,7 +57,7 @@ private[scala] trait PropertiesTrait {
   def setProp(name: String, value: String)      = System.setProperty(name, value)
   def clearProp(name: String)                   = System.clearProperty(name)
 
-  def envOrElse(name: String, alt: String)      = Option(System getenv name) getOrElse alt
+  def envOrElse(name: String, alt: => String)   = Option(System getenv name) getOrElse alt
   def envOrNone(name: String)                   = Option(System getenv name)
 
   def envOrSome(name: String, alt: Option[String])       = envOrNone(name) orElse alt
```
Comment (som-snytt) [inline src/library/scala/util/Properties.scala]: I think I suggested this before, but for completeness, convert all the API that correspond to Option, including `envOrSome`. And `scalaPropOrElse`?

Comment (193s) [inline src/library/scala/util/Properties.scala]: Oh, I didn't noticed `scalaPropOrElse`, thank you.

Comment (193s) [inline src/library/scala/util/Properties.scala]: fixed; https://github.com/193s/scala/commit/682fa115d5136d37bf19481b42cb5b293c3faf84

**Comment by som-snytt** [comment]
Interestingly, `test/files/run/t7775.scala` relies on the eager behavior for the test to test anything. (Though I didn't test that.)

I wonder if a test rig could have the feature that it can demonstrate that it is still testing something.

A proof that the predicate being asserted is still meaningful after all these years.

edit (by lrytz): the test is extended in #4372

**Comment by lrytz** [comment]
@adriaanm any ideas how we could push the change through (see your earlier comment)? i don't see a good way to keep the old signatures in the bytecode.

we could add overloads

```
propOrElse(name: String, alt: => String)
propOrElse(name: String, alt: String)
```

but then the compiler would pick the strict version, so the change would be pointless.

See #4297 for a similar problem.

@193s could you squash the three commits into one? (on your branch, run `git rebase -i 70f960cfd1`, then `git push -f`).

**Comment by 193s** [comment]
@lrytz I did as you said, thanks!

**Comment by adriaanm** [comment]
I think we can consider it a bug that the old *OrElse methods were eager, so we should fix that in 2.12. It seems unlikely that someone would expect eagerness here, or that these methods would be used in a performance hot spot. Let's add a deprecation message in 2.11 that says these methods will become call by name. The commit message of this PR should also explain why we made the change.

**Comment by lrytz** [comment]
@adriaanm agreed, but my question was about the testing infrastructure: how can we get a green build, given that partest links against the old signature?

@193s thanks for the squash! as suggested by Adriaan, could you provide a bit more detail in the commit message? I created a ticket (https://issues.scala-lang.org/browse/SI-9246), so you can refer to that. See here for a lengthy description about commit messages: https://github.com/scala/scala/wiki/Pull-Request-Policy

**Comment by 193s** [comment]
@lrytz, @adriaanm  Thanks for thanks for your kind advice.
I changed the commit message (e50fc2cd99677c5b1c1ff8746dc28177c7025e80), will that be ok?

**Comment by lrytz** [comment]
@193s looks good, thanks! Now you just have to be patient until we figure out how to get our test infrastructure working.

**Comment by adriaanm** [comment]
We'll have to release a new version of partest that uses reflection for one minor release I guess. Or that doesn't use the methods that break bin compat.

**Comment by som-snytt** [comment]
I bet no one would notice if you sucked the partest module back into scala during lunch hour today. There wouldn't be even the slightest dip in the community contribution rate, either.

**Comment by adriaanm** [comment]
True, but I'd rather not eat lunch than re-ingest that module. Completely agreed the partest burden should not be on the contributor, though.

**Comment by adriaanm** [comment]
Hi, once #4419 is merged, please `git rebase -i 2.11.x` and `git push --force 193s topic/prop2:topic/prop2`. This should allow this PR to go green, since partest no longer uses the *OrElse methods.

**Comment by adriaanm** [comment]
I've opened a new PR that does the rebase. I'll close this one. Thanks again for your contribution & patience ;-)

**Pull request closed without merging**